### PR TITLE
Account for page header in ORC output buffer

### PR DESCRIPTION
**Pull request by islamismailov**
```
== RELEASE NOTES ==

Hive Changes
* Add page header size accounting to ORC output buffer to avoid possible overflow.
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java [hunk 1]
```diff
@@ -444,13 +450,15 @@ private void writeChunkToOutputStream(byte[] chunk, int offset, int length)
 
         int compressedSize = compressor.compress(chunk, offset, length, compressionBuffer, 0, compressionBuffer.length);
         if (compressedSize < length) {
+            assert PAGE_HEADER_SIZE + compressedSize <= maxBufferSize;
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: The assert should be always on the length <= maxBufferSize. Checking on compressed size is incorrect as compressed size will always be less than length. maxBufferSize already accounts for PAGE_HEADER_SIZE and using that in addition to length will result in incorrect assertions.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java [hunk 2]
```diff
@@ -386,12 +388,14 @@ public String toString()
 
     private void ensureWritableBytes(int minWritableBytes)
     {
+        assert slice.length() <= maxBufferSize;
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: In my opinion this assert should be minWritableBytes <= maxBufferSize . sliceLength <= maxBufferSize will be verified in writeChunkToOutputStream.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java [hunk 3]
```diff
@@ -386,12 +388,14 @@ public String toString()
...
+
         int neededBufferSize = bufferPosition + minWritableBytes;
         if (neededBufferSize <= slice.length()) {
             return;
         }
 
-        if (slice.length() >= maxBufferSize) {
+        if (slice.length() == maxBufferSize) {
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: as defensive coding, this should be still >= and let the assert inside writeChunkToOutputStream fire.

Comment (islamismailov) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: This statement works with the assert above (`slice.length() <= maxBufferSize`), which makes sure that `maxBufferSize > slice.length()` is never the case, and we handle the only two other cases.

But I think your suggestions also make sense.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java (hunk 1 above)
Comment (islamismailov) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: Good catch. This is from the previous approach that I initially drafted.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java [hunk 4]
```diff
@@ -386,12 +387,14 @@ public String toString()
 
     private void ensureWritableBytes(int minWritableBytes)
     {
+        assert minWritableBytes <= maxBufferSize;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: We don't use `assert` in the codebase. Replace it with `checkState(...)` or `verify(...)` for state check or assertion. We also use  `checkArgument` to check arguments.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java [hunk 5]
```diff
@@ -49,6 +49,7 @@
         extends SliceOutput
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(OrcOutputBuffer.class).instanceSize();
+    private static final int PAGE_HEADER_SIZE = 3;
```
Comment (highker) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: Is this number `3` coming from the 3-bit header from ORC spec? If yes, could you add a comment?

**Comment by islamismailov** [comment]
Will update.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java (hunk 5 above)
Comment (islamismailov) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcOutputBuffer.java]: Yes, these are the 3 bytes that are used for the header.

https://github.com/apache/orc/blob/028261ad6e26d8ae9f9483c8aeb76a85c7f1168b/c%2B%2B/src/Compression.cc#L186

**Comment by islamismailov** [comment]
Please merge @highker / @wenleix  - as I do not have write access :). Thanks all!

**Pull request merged**